File: relationship-manager/conversation-service/src/main/java/com/bank/rm/conversation/ai/ConversationAIAgent.java
package com.bank.rm.conversation.ai;

import com.bank.rm.common.dto.ConversationPhase;
import dev.langchain4j.model.chat.ChatLanguageModel;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import java.util.Map;

public class ConversationAIAgent {
    private static final Logger log = LoggerFactory.getLogger(ConversationAIAgent.class);
    private final ChatLanguageModel chatModel;
    private final PromptManager promptManager;

    public ConversationAIAgent(ChatLanguageModel chatModel, PromptManager promptManager) {
        this.chatModel = chatModel;
        this.promptManager = promptManager;
    }

    public String generateResponse(String userMessage, ConversationPhase phase,
                                    Map<String, Object> extractedData, String conversationHistory) {
        try {
            String systemPrompt = promptManager.getSystemPrompt(phase);
            String fullPrompt = String.format(
                "System: %s\n\nConversation context:\n%s\n\nCollected data so far: %s\n\nCustomer: %s\n\nAssistant:",
                systemPrompt, conversationHistory, extractedData, userMessage);
            return chatModel.generate(fullPrompt);
        } catch (Exception e) {
            log.warn("LLM response generation failed, using template: {}", e.getMessage());
            return getTemplateFallback(phase, extractedData);
        }
    }

    public String generateSummary(String conversationHistory, Map<String, Object> extractedData) {
        try {
            String prompt = "Summarize this banking relationship manager conversation in 2-3 sentences:\n\n" +
                    conversationHistory + "\n\nCollected data: " + extractedData;
            return chatModel.generate(prompt);
        } catch (Exception e) {
            log.warn("Summary generation failed: {}", e.getMessage());
            return "Conversation completed. Data collected: " + extractedData.keySet();
        }
    }

    private String getTemplateFallback(ConversationPhase phase, Map<String, Object> data) {
        return switch (phase) {
            case GREETING -> "Welcome to our banking services! I'm your Relationship Manager. Could you please share your name?";
            case PERSONAL -> {
                String name = data.getOrDefault("name", "").toString();
                yield "Thank you" + (name.isEmpty() ? "" : ", " + name) +
                        "! Could you share your age group (20-30, 30-40, 40-50, or 50+), location, phone, and email?";
            }
            case FINANCIAL -> "Now, could you tell me about your income source and range, current savings, and investments?";
            case GOALS -> "What are your retirement goals? How much would you like to retire with, and by what age?";
            case RISK_ASSESSMENT -> "Based on your profile, I'll assess your risk tolerance. Would you describe yourself as conservative, moderate, or aggressive with investments?";
            case RECOMMENDATION -> "Based on your risk profile, here are my recommendations. I'll share the details shortly.";
            case CHANNEL_PREF -> "How would you like to stay in touch? SMS, WhatsApp, Email, or Phone?";
            case FOLLOWUP_SCHEDULE -> "Let's schedule regular check-ins. Would you prefer weekly, biweekly, monthly, or quarterly reviews?";
            case COMPLETED -> "Thank you for your time! Your profile is set up and we'll be in touch.";
        };
    }
}

File: relationship-manager/conversation-service/src/main/kotlin/com/bank/rm/conversation/ai/ConversationAIAgent.kt
package com.bank.rm.conversation.ai

import com.bank.rm.common.dto.ConversationPhase
import com.bank.rm.common.exception.AIServiceException
import dev.langchain4j.model.chat.ChatLanguageModel
import dev.langchain4j.model.input.Prompt
import dev.langchain4j.model.input.PromptTemplate
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component

@Component
class ConversationAIAgent(
    private val chatModel: ChatLanguageModel,
    private val promptManager: PromptManager,
    private val entityExtractor: EntityExtractor
) {
    private val log = LoggerFactory.getLogger(javaClass)

    fun generateResponse(
        phase: ConversationPhase,
        conversationHistory: List<ChatMessage>,
        extractedData: Map<String, Any>,
        customerMessage: String
    ): AIResponse {
        return try {
            val systemPrompt = promptManager.getSystemPrompt(phase, extractedData)
            val historyContext = conversationHistory.takeLast(10).joinToString("\n") {
                "${it.role}: ${it.content}"
            }

            val prompt = PromptTemplate.from(
                """{{system_prompt}}
                
Conversation so far:
{{history}}

Customer: {{message}}

Respond naturally as a banking relationship manager. Be empathetic, professional, and guide the conversation toward collecting the required information for this phase."""
            ).apply(
                mapOf(
                    "system_prompt" to systemPrompt,
                    "history" to historyContext,
                    "message" to customerMessage
                )
            )

            val response = chatModel.generate(prompt.text())
            val entities = entityExtractor.extract(customerMessage, phase)
            val shouldTransition = shouldTransitionPhase(phase, extractedData + entities)

            AIResponse(
                message = response,
                extractedEntities = entities,
                shouldTransitionPhase = shouldTransition,
                suggestedNextPhase = if (shouldTransition) getNextPhase(phase) else null
            )
        } catch (e: Exception) {
            log.error("AI agent error for phase $phase: ${e.message}", e)
            throw AIServiceException("Failed to generate AI response: ${e.message}", e)
        }
    }

    fun generateSummary(conversationHistory: List<ChatMessage>, extractedData: Map<String, Any>): String {
        return try {
            val prompt = PromptTemplate.from(
                """Summarize this banking relationship conversation concisely. Include:
- Customer profile highlights
- Key financial details discussed
- Risk assessment outcome
- Products recommended
- Follow-up actions agreed

Conversation:
{{history}}

Extracted data:
{{data}}

Summary:"""
            ).apply(
                mapOf(
                    "history" to conversationHistory.joinToString("\n") { "${it.role}: ${it.content}" },
                    "data" to extractedData.toString()
                )
            )
            chatModel.generate(prompt.text())
        } catch (e: Exception) {
            log.warn("Failed to generate AI summary, using template fallback", e)
            generateTemplateSummary(extractedData)
        }
    }

    fun detectHandoffIntent(message: String): Boolean {
        return try {
            val response = chatModel.generate(
                """Analyze if this customer message indicates they want to speak with a human agent.
Reply with only "YES" or "NO".
Message: "$message""""
            )
            response.trim().uppercase().startsWith("YES")
        } catch (e: Exception) {
            log.warn("Handoff detection failed, checking keywords", e)
            val keywords = listOf("human", "agent", "person", "manager", "real person", "speak to someone")
            keywords.any { message.lowercase().contains(it) }
        }
    }

    private fun shouldTransitionPhase(phase: ConversationPhase, data: Map<String, Any>): Boolean {
        val requiredFields = when (phase) {
            ConversationPhase.GREETING -> listOf("acknowledged")
            ConversationPhase.PERSONAL -> listOf("name", "ageGroup")
            ConversationPhase.FINANCIAL -> listOf("incomeSource", "incomeRange")
            ConversationPhase.GOALS -> listOf("retirementTarget")
            ConversationPhase.RISK_ASSESSMENT -> listOf("riskScore")
            ConversationPhase.RECOMMENDATION -> listOf("recommendationsPresented")
            ConversationPhase.CHANNEL_PREF -> listOf("preferredChannel")
            ConversationPhase.FOLLOWUP_SCHEDULE -> listOf("followUpFrequency")
            ConversationPhase.COMPLETED -> emptyList()
        }
        return requiredFields.all { data.containsKey(it) }
    }

    private fun getNextPhase(current: ConversationPhase): ConversationPhase = when (current) {
        ConversationPhase.GREETING -> ConversationPhase.PERSONAL
        ConversationPhase.PERSONAL -> ConversationPhase.FINANCIAL
        ConversationPhase.FINANCIAL -> ConversationPhase.GOALS
        ConversationPhase.GOALS -> ConversationPhase.RISK_ASSESSMENT
        ConversationPhase.RISK_ASSESSMENT -> ConversationPhase.RECOMMENDATION
        ConversationPhase.RECOMMENDATION -> ConversationPhase.CHANNEL_PREF
        ConversationPhase.CHANNEL_PREF -> ConversationPhase.FOLLOWUP_SCHEDULE
        ConversationPhase.FOLLOWUP_SCHEDULE -> ConversationPhase.COMPLETED
        ConversationPhase.COMPLETED -> ConversationPhase.COMPLETED
    }

    private fun generateTemplateSummary(data: Map<String, Any>): String {
        return buildString {
            append("Conversation Summary:\n")
            data["name"]?.let { append("- Customer: $it\n") }
            data["ageGroup"]?.let { append("- Age Group: $it\n") }
            data["incomeSource"]?.let { append("- Income: $it\n") }
            data["riskCategory"]?.let { append("- Risk Profile: $it\n") }
            data["preferredChannel"]?.let { append("- Preferred Contact: $it\n") }
        }
    }
}

data class AIResponse(
    val message: String,
    val extractedEntities: Map<String, Any>,
    val shouldTransitionPhase: Boolean,
    val suggestedNextPhase: ConversationPhase?
)

data class ChatMessage(
    val role: String,
    val content: String
)
